### ISSUE TITLE
Make passing options more convenient

### ISSUE THREAD

**Issue by FiV0** [cnuernber/charred#13]
Currently when one needs to pass options down to charred this can lead to combinatorial explosion.
```clj
(cond
  (and key-fn return-type) #(charred/read-json % :key-fn key-fn :value-fn return-type)
  key-fn #(charred/read-json % :key-fn key-fn)
  return-type #(charred/read-json % :value-fn return-type)
  :else charred/read-json)
```
Would you consider making this possible via a map which you are constructing here anyway: https://github.com/cnuernber/charred/blob/b081dc23ce9174dc53dcd983f98b7e4815de3344/src/charred/api.clj#L531
You could also pin to 1.11 which supports this out of the box https://clojure.org/news/2021/03/18/apis-serving-people-and-programs.

**Comment by cnuernber** [comment]
This has been an irritation of mine with clojure `& options` style functions for literally the last 12 years.

This is exactly why, when I am not trying to make a drop-in replacement for clojure.data.csv or data.json I use an explicit options map. 

You can choose to use 1.11 - I don't see why I would pin to it.  

Why not simply use the `-supplier` pathways yourself?

**Comment by FiV0** [comment]
What I meant (using clojure 1.11):
```clj
(require '[clojure.data.json :as json])

(json/read-str "{\"a\":1,\"b\":2}" {:key-fn  keyword})
;; => {:a 1, :b 2}

(require '[charred.api :as charred])

(charred/read-json "{\"a\":1,\"b\":2}" {:key-fn  keyword})
;; => {"a" 1, "b" 2}
```
replacing `read-json` with
```clj
(defn read-json
  "Drop in replacement for clojure.data.json/read and clojure.data.json/read-str.  For options
  see [[read-json-supplier]]."
  [input & {:as args}]
  (with-open [json-fn (read-json-supplier input args)]
    (.get json-fn)))
```
results in 
```clj 
(charred/read-json "{\"a\":1,\"b\":2}" :key-fn  keyword)
;; => {:a 1, :b 2}

(charred/read-json "{\"a\":1,\"b\":2}" {:key-fn  keyword})
;; => {:a 1, :b 2}
```
so you don't break downstream and make it possible to have option maps with 1.11.

**Comment by cnuernber** [comment]
Yes this is totally reasonable.  Fixed 1.015 for all functions that had that style of options passing.
